beacon-chain/sync/backfill/status.go

```diff
@@ -2,121 +2,169 @@ package backfill
 		return true
 	}
-	if s.StartGap() < sl && sl < s.EndGap() {
-		return false
-	}
-	return true
-}
-
-// StartGap returns the slot at the beginning of the range that needs to be backfilled.
-func (s *Status) StartGap() primitives.Slot {
-	return s.start
+	return false
 }
 
-// EndGap returns the slot at the end of the range that needs to be backfilled.
-func (s *Status) EndGap() primitives.Slot {
-	return s.end
+// Status is a threadsafe method to access a copy of the BackfillStatus value.
+func (s *StatusUpdater) status() *dbval.BackfillStatus {
+	s.RLock()
+	defer s.RUnlock()
+	return &dbval.BackfillStatus{
+		LowSlot:       s.bs.LowSlot,
+		LowRoot:       s.bs.LowRoot,
+		LowParentRoot: s.bs.LowParentRoot,
+		OriginSlot:    s.bs.OriginSlot,
+		OriginRoot:    s.bs.OriginRoot,
+	}
 }
 
-var ErrAdvancePastOrigin = errors.New("cannot advance backfill Status beyond the origin checkpoint slot")
+// fillBack saves the slice of blocks and updates the BackfillStatus LowSlot/Root/ParentRoot tracker to the values
+// from the first block in the slice. This method assumes that the block slice has been fully validated and
+// sorted in slot order by the calling function.
+func (s *StatusUpdater) fillBack(ctx context.Context, blocks []blocks.ROBlock) (*dbval.BackfillStatus, error) {
+	status := s.status()
+	if len(blocks) == 0 {
+		return status, nil
+	}
 
-// Advance advances the backfill position to the given slot & root.
-// It updates the backfill block root entry in the database,
-// and also updates the Status value's copy of the backfill position slot.
-func (s *Status) Advance(ctx context.Context, upTo primitives.Slot, root [32]byte) error {
-	if upTo > s.end {
-		return errors.Wrapf(ErrAdvancePastOrigin, "advance slot=%d, origin slot=%d", upTo, s.end)
+	highest := blocks[len(blocks)-1]
+	// The root of the highest block needs to match the parent root of the previous status. The backfill service will do
+	// the same check, but this is an extra defensive layer in front of the db index.
+	if highest.Root() != bytesutil.ToBytes32(status.LowParentRoot) {
+		return nil, errors.Wrapf(errBatchDisconnected, "prev parent_root=%#x, root=%#x, prev slot=%d, slot=%d",
+			status.LowParentRoot, highest.Root(), status.LowSlot, highest.Block().Slot())
 	}
-	s.start = upTo
-	return s.store.SaveBackfillBlockRoot(ctx, root)
-}
 
-// Reload queries the database for backfill status, initializing the internal data and validating the database state.
-func (s *Status) Reload(ctx context.Context) error {
-	cpRoot, err := s.store.OriginCheckpointBlockRoot(ctx)
-	if err != nil {
-		// mark genesis sync and short circuit further lookups
-		if errors.Is(err, db.ErrNotFoundOriginBlockRoot) {
-			s.genesisSync = true
-			return nil
+	for _, b := range blocks {
+		if err := s.store.SaveBlock(ctx, b); err != nil {
+			return nil, errors.Wrapf(err, "error saving backfill block with root=%#x, slot=%d", b.Root(), b.Block().Slot())
```

Comment:
Nit. I found that `StatusUpdater` not only updates statuses but also handles DB block saving and finalized root updates. This seems to diverge from the principle of single responsibility based on the name. Maybe renaming a few things fix this?

Reply:
I renamed the type to `Store` to reflect the fact that it exposes a broader swath of the db interface. What do you think?